cachi2/core/package_managers/gomod.py

```diff
@@ -314,98 +483,30 @@ def go_list_deps(pattern: Literal["./...", "all"]) -> Iterator[GoPackage]:
-        chain(
-            map(_get_name_and_version, downloaded_modules),
-            # package_modules should generally be a subset of downloaded_modules,
-            # but modules replaced with local paths may only appear in package_modules
-            map(_get_name_and_version, package_modules),
-        )
-    )
+        return map(ParsedPackage.parse_obj, load_json_stream(_run_gomod_cmd(cmd, run_params)))
 
-    main_module_deps = [
-        {"type": "gomod", "name": name, "version": version}
-        for name, version in sorted(module_names_and_versions)
-    ]
+    package_modules = [mod for pkg in go_list_deps("all") if (mod := pkg.module) and not mod.main]
+    module_deps = downloaded_modules + package_modules
 
     log.info("Retrieving the list of packages")
-    main_packages: list[dict[str, Any]] = []
-    main_packages_deps = set()
-    all_packages = {pkg.import_path: pkg for pkg in go_list_deps("./...")}
-
-    # The go list -deps command lists dependencies first and then the package which
-    # depends on them. We want the opposite order: package first, then its dependencies.
-    for pkg in reversed(all_packages.values()):
-        if not pkg.module or not pkg.module.main:
-            continue  # This is a dependency, not a top-level package
-
-        if pkg.import_path in main_packages_deps:
-            # If a top-level package is already listed as a dependency, we do not list it here,
-            # since its dependencies are already listed in the parent package.
-            log.debug(
-                "Package %s is already listed as a package dependency. Skipping...",
-                pkg.import_path,
-            )
-            continue
+    all_packages = [pkg for pkg in go_list_deps("./...")]
+
+    _validate_local_replacements(module_deps, app_dir)
+    
+    # why do we need to do it for pkgs?
+    # _vet_local_deps(all_packages)
 
-        pkg_deps = []
-        for dep_name in pkg.deps:
-            dep = all_packages[dep_name]
-            main_packages_deps.add(dep_name)
-
-            if dep.standard:
-                dep_version = None
-            elif not dep.module or dep.module.main:
-                # Standard=false, Module.Main=true
-                # Standard=false, Module=null   <- probably a to-be-generated package
-                dep_version = main_module_version
-            else:
-                _, dep_version = _get_name_and_version(dep.module)
-
-            pkg_deps.append({"type": "go-package", "name": dep_name, "version": dep_version})
-
-        main_pkg = {
-            "type": "go-package",
-            "name": pkg.import_path,
-            "version": main_module_version,
-        }
-        main_packages.append({"pkg": main_pkg, "pkg_deps": pkg_deps})
-
-    _vet_local_deps(main_module_deps)
-    for package in main_packages:
-        # Local dependencies are always relative to the main module, even for subpackages
-        _vet_local_deps(package["pkg_deps"])
```

Comment:
Why did we need to check for replacements of packages? They can't be replaced, right?

Reply:
Hmm, good question. Maybe in Cachito, all the package names had to match the allowlist pattern? But now that there's no allolist, it should be 100% useless

Reply:
Well, I removed this bit and kept the check only on module level.